### PR TITLE
Oauth

### LINKED ISSUE #670: Build Twitter promo
image

### PR DESCRIPTION
**Pull request by davidpmccormick**
Closes #670. References #668.

## Type
✨ Feature  

## Value
Adding the four most recent tweets from @explorewellcome to the page using the Twitter API and our async component.

## Screenshot
screen shot 2017-06-02 at 17 25 53

## TODO
~- [ ] Abstract component to allow for Instagram/other (as well as just Twitter)~ <-- will do this work in a separate PR.

### All
- [ ] Demoed to the relevant people
- [x] PR labelled and assigned
- [x] Any introduced code complexity has been flagged

### UI component only
- [x] A11y tested: `npm run test:accessibility <URL>`
- [x] Browser tested: `npm run test:browsers <URL>`
- [x] Works without JS in the client
- [x] Relevant tracking/monitoring has been considered

### CODE REVIEW

File: server/controllers/index.js [hunk 1]
```diff
@@ -144,6 +145,19 @@ export const seriesTransporter = async(ctx, next) => {
...
 
+export const latestTweets = async(ctx, next) => {
+  const count = ctx.params.count || 4;
+  const tweets = await getLatestTweets(count);
+
+  ctx.render('components/tweet-block/index', {
+    model: { tweets }
+  });
+
+  ctx.body = {
```
Comment (jamesgorrie) [inline server/controllers/index.js]: Why not use render?

File: server/services/twitter.js [hunk 2]
```diff
@@ -0,0 +1,62 @@
+import superagent from 'superagent';
+import { createTweet } from '../model/tweet';
+
+async function getTwitterAccessToken() {
+  // TODO: reimplement after testing
+  const urlEncodedK = encodeURIComponent('qpsVJlsTg22FFKr1fwzkzU4Ew');
+  const urlEncodedS = encodeURIComponent('kmXPnluTYkl7oH4mRlC1EgthH8KLT9mBpezgK3wBaFQviC23Ru');
```
Comment (jamesgorrie) [inline server/services/twitter.js]: Are we 100% sure these don't have write access?

File: server/services/twitter.js [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+  const base64EncodedKS = Buffer.from(`${urlEncodedK}:${urlEncodedS}`).toString('base64');
+
+  return new Promise((resolve, reject) => {
```
Comment (jamesgorrie) [inline server/services/twitter.js]: Do we need to use this?  We could use `async` `await` as `superagent` is already Promisified.

File: server/services/twitter.js [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+        resolve(res.body.access_token);
+      });
+  });
+}
+
+async function getLatestTweetDetails(count) {
+  const accessToken = await getTwitterAccessToken();
+
+  return new Promise((resolve, reject) => {
```
Comment (jamesgorrie) [inline server/services/twitter.js]: Ditto as above.

File: server/services/twitter.js [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    </blockquote>`;
+
+    return createTweet({
+      id: details.id,
+      screenName: details.screenName,
+      createdAt: details.createdAt,
+      html: embedTweetHtml
+    });
+  });
+}
```
Comment (jamesgorrie) [inline server/services/twitter.js]: Yea, we could definitely use async await for all of this - it'd make it a bit tidier.

File: server/views/layout/default.njk [hunk 6]
```diff
@@ -19,6 +19,7 @@
     <link rel="manifest" href="/assets/icons/manifest.json"/>
     <link rel="mask-icon" href="/assets/icons/safari-pinned-tab.svg" color="#000000"/>
     <script src="/assets/libs/picturefill.min.js" async></script>
+    <script async defer src="//platform.twitter.com/widgets.js"></script>
```
Comment (jamesgorrie) [inline server/views/layout/default.njk]: We could have a race condition here where this doesn't load before we run the code above...
Happy for now until we see an error.

What % of the browsers we support support async / defer?

File: server/views/pages/explore.njk [hunk 7]
```diff
@@ -75,5 +75,9 @@
     </div>
   </div>
 
+  <div data-component-name="Tweet block" data-component-id="explore">
```
Comment (jamesgorrie) [inline server/views/pages/explore.njk]: `component-id="explore"`?

File: server/views/layout/default.njk (hunk 6 above)
Comment (davidpmccormick) [inline server/views/layout/default.njk]: `async` and `defer` are in all of our supported browsers, impressively.

File: server/views/pages/explore.njk (hunk 7 above)
Comment (davidpmccormick) [inline server/views/pages/explore.njk]: As in, 'The "Tweet block" that appears on the explore page' – wasn't sure if there's a better way of differentiating between them. I'm open to suggestions.

File: server/services/twitter.js (hunk 5 above)
Comment (davidpmccormick) [inline server/services/twitter.js]: Ah yeah – will update.

File: server/services/twitter.js (hunk 2 above)
Comment (davidpmccormick) [inline server/services/twitter.js]: Yeah, we're OK

screen shot 2017-06-05 at 11 36 16

File: server/views/pages/explore.njk (hunk 7 above)
Comment (jamesgorrie) [inline server/views/pages/explore.njk]: I'm still a little confused about what `name` vs `id` is used for?
`name` for tracking, id for ?

Comment (davidpmccormick) [inline server/views/pages/explore.njk]: `id` for when we have more than one of the type of component on the site

Comment (jamesgorrie) [inline server/views/pages/explore.njk]: Ah - got it - Okay, I'll perhaps document this somewhere, as it's a little confusing (for me).

File: server/controllers/index.js [hunk 8]
```diff
@@ -144,6 +145,19 @@ export const seriesTransporter = async(ctx, next) => {
   return next();
 };
 
+export const latestTweets = async(ctx, next) => {
```
Comment (jamesgorrie) [inline server/controllers/index.js]: I wonder if we should leave a note, or potentially break this out into something else (`async-controllers`?) to make sure people know not to use these in the sync world due to the massive round trip... just a thought, happy not to for now.

**Pull request merged**